rmw_zenoh_cpp/src/rmw_service.cpp

```diff
@@ -0,0 +1,419 @@
+  RMW_CHECK_ARGUMENT_FOR_NULL(node, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(service, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    node,
+    node->implementation_identifier,
+    eclipse_zenoh_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    service,
+    service->implementation_identifier,
+    eclipse_zenoh_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+
+  // OBTAIN ALLOCATOR ==========================================================
+  rcutils_allocator_t * allocator = &node->context->options.allocator;
+
+  // CLEANUP ===================================================================
+  auto service_data = static_cast<rmw_service_data_t *>(service->data);
+
+  zn_undeclare_subscriber(service_data->zn_request_subscriber_);
+  zn_undeclare_queryable(service_data->zn_queryable_);
+
+  allocator->deallocate(service_data->request_type_support_,
+                        allocator->state);
+  allocator->deallocate(service_data->response_type_support_,
+                        allocator->state);
+  allocator->deallocate(service->data, allocator->state);
+  allocator->deallocate(service, allocator->state);
+
+  return RMW_RET_OK;
+}
+
+/// TAKE REQUEST MESSAGE =======================================================
+// Take request message out of the request message queue
+rmw_ret_t
+rmw_take_request(
+  const rmw_service_t * service,
+  rmw_service_info_t * request_header,  // Out parameter
+  void * ros_request,
+  bool * taken)
+{
+  *taken = false;
+  // RCUTILS_LOG_INFO_NAMED("rmw_zenoh_cpp", "rmw_take_request");
+
+  // ASSERTIONS ================================================================
+  RMW_CHECK_ARGUMENT_FOR_NULL(service, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(ros_request, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(taken, RMW_RET_INVALID_ARGUMENT);
+
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    service,
+    service->implementation_identifier,
+    eclipse_zenoh_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION
+  );
+
+  // OBTAIN SUBSCRIPTION MEMBERS ===============================================
+  const char * service_name = service->service_name;
+  RMW_CHECK_ARGUMENT_FOR_NULL(service_name, RMW_RET_INVALID_ARGUMENT);
+
+  rmw_service_data_t * service_data = static_cast<rmw_service_data_t *>(service->data);
+  RMW_CHECK_ARGUMENT_FOR_NULL(service_data, RMW_RET_ERROR);
+
+  // OBTAIN ALLOCATOR ==========================================================
+  rcutils_allocator_t * allocator = &service_data->node_->context->options.allocator;
+
+  // RETRIEVE SERIALIZED MESSAGE ===============================================
+  std::string key(service_data->zn_request_topic_key_);
+
+  if (service_data->zn_request_messages_.find(key) == service_data->zn_request_messages_.end()) {
+    return RMW_RET_OK;
```

Comment:
This should probably be an error, shouldn't it?

Reply:
As above, it's not an error.

64c646b21efe62d75ae1089f8ded2e6f9378512d